navigation/image_database.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+// BoB robotics includes
+#include "../common/pose.h"
+
+// Third-party includes
+#include "../third_party/path.h"
+#include "../third_party/units.h"
+
+// OpenCV includes
+#include <opencv2/opencv.hpp>
+
+// Standard C includes
+#include <cassert>
+#include <cstdio>
+
+// Standard C++ includes
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace BoBRobotics {
+namespace Navigation {
+using namespace units::length;
+using namespace units::angle;
+
+class ImageDatabase
+{
+public:
+    struct Entry {
+        Vector3<millimeter_t> position;
+        degree_t heading;
+        std::string path;
+
+        cv::Mat load() const
+        {
+            return cv::imread(path);
+        }
+
+        cv::Mat loadGreyscale() const
+        {
+            return cv::imread(path, cv::IMREAD_GRAYSCALE);;
+        }
+    };
+
+    template<class T>
+    ImageDatabase(T &&databaseName)
```

Comment:
Is this templating really necessary? Can't you just take a const reference to a path and assume that filesystem::path provides sensible implicit constructors?

Reply:
Yep, we could do. But I put the templating in so that one can do:
```
void myfunc(const ImageDatabase &db) { ... }

myfunc("mydatabase/");
```

I could/should just have string + const char * ctors for this though to avoid excessive templating (which is one of my many vices!)